### PR TITLE
Traps can directly trigger EoCs as their action

### PR DESCRIPTION
**Pull request by Procyonae**
#### Summary
Infrastructure "Traps can directly trigger EoCs as their action"

#### Purpose of change
Triggering simple EoCs via spells is ugly and cumbersome

#### Describe the solution
Traps can trigger EoCs directly with the creature that triggers it as the alpha talker
Passes the trap's location as a context variable `trap_location` for use primarily with ranged traps
~~If triggered by an item passes the item's id as a context variable `trap_item_trigger`, so you can do cool silly stuff like puzzle switches where putting different key items on a trap does different things and you need to figure out the solution, with incorrect placement potentially having bad consequences (Erk posited something like this for nether dungeons before)~~
Documents the above
Unhardcodes the trap action "telepad" (and "portal" which was just an alias) bc I needed to write an EoC to test it anyway
Changes trap->spell->eoc to trap->eoc across the repo

#### Describe alternatives you've considered
Doing this in #77455 but it's not required for those changes and could do with a repo wide sweep up of existing pseudo EoC traps
I attempted items that trigger traps to be passed as talkers but couldn't get that to work, I considered falling back on passing a `trap_item_trigger` context var with the id of the item (and maybe another for it's translated name) but that would require passing an irrelevant talker like the player as you can't have 0 talker dialogues which seems very hacky. Will attempt to try this again if I think of anything

#### Testing
The tr_telepad works as expected with the player and Liam

#### Additional context
Firing EoCs via spells can still make sense if you want something complicated

### CODE REVIEW

File: src/trap.cpp [hunk 1]
```diff
@@ -155,7 +155,22 @@ void trap::load( const JsonObject &jo, const std::string_view )
     int legacy_floor_bedding_warmth = units::to_legacy_bodypart_temp_delta( floor_bedding_warmth );
     optional( jo, was_loaded, "floor_bedding_warmth", legacy_floor_bedding_warmth, 0 );
     floor_bedding_warmth = units::from_legacy_bodypart_temp_delta( legacy_floor_bedding_warmth );
-    optional( jo, was_loaded, "spell_data", spell_data );
+    if( jo.has_member( "spell_data" ) ) {
+        //This is kinda ugly but idk how to do it better bc std::function doesn't support normal equality
+        if( act.target_type() != trap_function_from_string( "spell" ).target_type() ) {
+            jo.throw_error_at( "spell_data",
+                               "Can't use \"spell_data\" without specifying \"action\": \"spell\"" );
```
Comment (Procyonae) [inline src/trap.cpp:162]:
```suggestion
                               R"(Can't use "spell_data" without specifying "action": "spell")" );
```

File: src/trap.cpp [hunk 2]
```diff
@@ -155,7 +155,22 @@ void trap::load( const JsonObject &jo, const std::string_view )
...
+        }
+        optional( jo, was_loaded, "spell_data", spell_data );
+    }
+    if( jo.has_member( "eocs" ) ) {
+        if( act.target_type() != trap_function_from_string( "eocs" ).target_type() ) {
+            jo.throw_error_at( "eocs", "Can't use \"eocs\" without specifying \"action\": \"eocs\"" );
```
Comment (Procyonae) [inline src/trap.cpp:168]:
```suggestion
            jo.throw_error_at( "eocs", R"(Can't use "eocs" without specifying "action": "eocs")" );
```

**Comment by Night-Pryanik** [comment]
@Procyonae could you please resolve conflicts?

**Comment by Procyonae** [comment]
Waiting on #77808 merging first (and were waiting on Guardian's big EoC PR before that)

**Pull request merged**